File: system/backend/app/services/tag_service.py
"""
Member tagging service
Manages tags/categories for system members
"""

import json
from typing import List, Dict
from app.core.config import MEMBER_TAGS_FILE

# Default member tag assignments
DEFAULT_MEMBER_TAGS = {
    "Jinx": ["Arcane"],
    "Vi": ["Arcane"],
    "Onyx": ["Cat"],
    "Pyzer": ["Cat"],
    "Pyzen": ["Cat"],
    "JS": ["Cyberpunk"],
    "V": ["Cyberpunk"],
    "Judy": ["Cyberpunk"],
    "Marin": ["DUD"],
    "Roxy": ["FNAF"],
    "Catrin": ["Fortnite"],
    "Hope": ["Fortnite"],
    "Meg": ["Fortnite"],
    "02": ["Franxx"],
    "C1": ["Host"],
    "Cleo": ["Host"],
    "Cwove": ["Host"],
    "abby": ["KPDH"],
    "baby": ["KPDH"],
    "Bobby": ["KPDH"],
    "jinu": ["KPDH"],
    "Mira": ["KPDH"],
    "mystery": ["KPDH"],
    "romance": ["KPDH"],
    "Rumi": ["KPDH"],
    "Zoey": ["KPDH"],
    "Dashie": ["MLP"],
    "D.Va": ["Overwatch"],
    "Tracer": ["Overwatch"],
    "Astra": ["Valorant"],
    "C2": ["Valorant"],
    "Cypher": ["Valorant"],
    "Deadlock": ["Valorant"],
    "Fade": ["Valorant"],
    "Jett": ["Valorant"],
    "KJ": ["Valorant"],
    "Neon": ["Valorant"],
    "Raze": ["Valorant"],
    "Reyna": ["Valorant"],
    "Sage": ["Valorant"],
    "Viper": ["Valorant"],
    "Vyse": ["Valorant"],
    "CMiku": ["Vocaloids"],
    "HMiku": ["Vocaloids"],
    "SMiku": ["Vocaloids"],
    "Anby": ["ZZZ"],
    "Belle": ["ZZZ"]
}

def get_member_tags() -> Dict[str, List[str]]:
    """Get all member tag assignments"""
    if not MEMBER_TAGS_FILE.exists():
        save_member_tags(DEFAULT_MEMBER_TAGS)
        return DEFAULT_MEMBER_TAGS.copy()
    
    with open(MEMBER_TAGS_FILE, "r") as f:
        return json.load(f)

def save_member_tags(member_tags: Dict[str, List[str]]):
    """Save member tags to file"""
    with open(MEMBER_TAGS_FILE, "w") as f:
        json.dump(member_tags, f, indent=2)

def get_member_tags_by_id(member_id: str, member_name: str) -> List[str]:
    """Get tags for a specific member by ID or name"""
    member_tags = get_member_tags()
    
    # Try by member name first
    if member_name in member_tags:
        return member_tags[member_name]
    
    # Then try by member ID
    if member_id in member_tags:
        return member_tags[member_id]
    
    return []

def update_member_tags(member_identifier: str, tags: List[str]) -> bool:
    """Update tags for a member (can use ID or name)"""
    member_tags = get_member_tags()
    member_tags[member_identifier] = tags
    save_member_tags(member_tags)
    return True

def add_member_tag(member_identifier: str, tag: str) -> bool:
    """Add a single tag to a member"""
    member_tags = get_member_tags()
    if member_identifier not in member_tags:
        member_tags[member_identifier] = []
    
    if tag not in member_tags[member_identifier]:
        member_tags[member_identifier].append(tag)
        save_member_tags(member_tags)
        return True
    
    return False

def remove_member_tag(member_identifier: str, tag: str) -> bool:
    """Remove a single tag from a member"""
    member_tags = get_member_tags()
    if member_identifier in member_tags and tag in member_tags[member_identifier]:
        member_tags[member_identifier].remove(tag)
        save_member_tags(member_tags)
        return True
    
    return False

def enrich_members_with_tags(members: List[Dict]) -> List[Dict]:
    """Add tag information to all members"""
    enriched_members = []
    
    for member in members:
        member_name = member.get("name", "")
        member_id = member.get("id", "")
        
        # Get tags for this member
        tags = get_member_tags_by_id(member_id, member_name)
        
        # Add tags to member data
        member_with_tags = {**member, "tags": tags}
        enriched_members.append(member_with_tags)
    
    return enriched_members

def initialize_default_tags():
    """Initialize default member tags if they don't exist"""
    if not MEMBER_TAGS_FILE.exists():
        save_member_tags(DEFAULT_MEMBER_TAGS)
        print("Initialized default member tags")

File: system/backend/src/services/tag_service.ts
/**
 * Copyright (c) 2026 Clove Twilight
 * Licensed under the ESAL-2.0 Licence.
 * See LICENCE.md in the project root for full licence information.
 */

/**
 * Member tagging service.
 * Manages tags/categories for system members.
 */

import { existsSync } from 'node:fs';
import { readFile, writeFile } from 'node:fs/promises';

import { MEMBER_TAGS_FILE } from '../core/config.js';

// Generic PluralKit-shaped member object, same rationale as pluralkit.ts
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type PKObject = Record<string, any>;

type MemberTagsMap = Record<string, string[]>;

// Default member tag assignments
const DEFAULT_MEMBER_TAGS: MemberTagsMap = {
  C1: ['Host'],
};

/** Get all member tag assignments */
export async function getMemberTags(): Promise<MemberTagsMap> {
  if (!existsSync(MEMBER_TAGS_FILE)) {
    await saveMemberTags(DEFAULT_MEMBER_TAGS);
    return { ...DEFAULT_MEMBER_TAGS };
  }

  const raw = await readFile(MEMBER_TAGS_FILE, 'utf-8');
  return JSON.parse(raw) as MemberTagsMap;
}

/** Save member tags to file */
export async function saveMemberTags(memberTags: MemberTagsMap): Promise<void> {
  await writeFile(MEMBER_TAGS_FILE, JSON.stringify(memberTags, null, 2), 'utf-8');
}

/** Get tags for a specific member by ID or name */
export async function getMemberTagsById(memberId: string, memberName: string): Promise<string[]> {
  const memberTags = await getMemberTags();

  // Try by member name first
  if (memberName in memberTags) {
    return memberTags[memberName];
  }

  // Then try by member ID
  if (memberId in memberTags) {
    return memberTags[memberId];
  }

  return [];
}

/** Update tags for a member (can use ID or name) */
export async function updateMemberTags(memberIdentifier: string, tags: string[]): Promise<boolean> {
  const memberTags = await getMemberTags();
  memberTags[memberIdentifier] = tags;
  await saveMemberTags(memberTags);
  return true;
}

/** Add a single tag to a member */
export async function addMemberTag(memberIdentifier: string, tag: string): Promise<boolean> {
  const memberTags = await getMemberTags();
  if (!(memberIdentifier in memberTags)) {
    memberTags[memberIdentifier] = [];
  }

  if (!memberTags[memberIdentifier].includes(tag)) {
    memberTags[memberIdentifier].push(tag);
    await saveMemberTags(memberTags);
    return true;
  }

  return false;
}

/** Remove a single tag from a member */
export async function removeMemberTag(memberIdentifier: string, tag: string): Promise<boolean> {
  const memberTags = await getMemberTags();
  const tags = memberTags[memberIdentifier];

  if (tags && tags.includes(tag)) {
    memberTags[memberIdentifier] = tags.filter((t) => t !== tag);
    await saveMemberTags(memberTags);
    return true;
  }

  return false;
}

/** Add tag information to all members */
export async function enrichMembersWithTags(members: PKObject[]): Promise<PKObject[]> {
  const enrichedMembers: PKObject[] = [];

  for (const member of members) {
    const memberName = member.name ?? '';
    const memberId = member.id ?? '';

    const tags = await getMemberTagsById(memberId, memberName);

    enrichedMembers.push({ ...member, tags });
  }

  return enrichedMembers;
}

/** Initialize default member tags if they don't exist */
export async function initializeDefaultTags(): Promise<void> {
  if (!existsSync(MEMBER_TAGS_FILE)) {
    await saveMemberTags(DEFAULT_MEMBER_TAGS);
    console.info('Initialized default member tags');
  }
}